air/src/execution_step/instructions/mod.rs

```diff
@@ -46,28 +46,18 @@ use super::ExecutionResult;
 use crate::execution_step::TraceHandler;
 
 use air_parser::ast::Instruction;
-
-/// Executes an instruction and updates both %last_error% and :error: if necessary.
-/// The difference b/w the two following macroses is that the first additionaly stores
-/// peer_id in the errors whilst the second does not.
-macro_rules! execute_both_errors_w_peerid {
-    ($self:expr, $instr:expr, $exec_ctx:ident, $trace_ctx:ident) => {{
-        match $instr.execute($exec_ctx, $trace_ctx) {
-            Err(e) => {
-                $exec_ctx.set_errors_w_peerid(&e, &$instr.to_string(), None);
-                Err(e)
-            }
-            v => v,
-        }
-    }};
-}
+use air_parser::ast::PeerIDErrorLogable;
 
 /// Executes an instruction and updates %last_error% and :error: if necessary.
 macro_rules! execute {
     ($self:expr, $instr:expr, $exec_ctx:ident, $trace_ctx:ident) => {{
         match $instr.execute($exec_ctx, $trace_ctx) {
             Err(e) => {
-                $exec_ctx.set_errors(&e, &$instr.to_string(), None);
+                if $instr.log_errors_with_peer_id() {
```

Comment:
Every invocation of `execute!` will have this branching.  Instead, we might pass the boolean flag to single function `set_errors`.

Reply:
I think it's better to pass peer_id. `set_errors` should be 

```rust
    pub(crate) fn set_errors(
        &mut self,
        error: &(impl ErrorEffectable + ToErrorCode + ToString),
        peer_id: &String,
        instruction: &str,
        tetraplet: Option<RcSecurityTetraplet>,
    ) {
        self.last_error_descriptor.try_to_set_last_error_from_exec_error(
            error,
            instruction,
            Some(peer_id),
            tetraplet.clone(),
        );
        self.error_descriptor
            .try_to_set_error_from_exec_error(error, instruction, None, tetraplet);
    }
```

This macro should have similar to this view:
```rust
macro_rules! execute {
    ($self:expr, $instr:expr, $exec_ctx:ident, $trace_ctx:ident) => {{
        match $instr.execute($exec_ctx, $trace_ctx) {
            Err(e) => {
                let peer_id = if let Some(peer_id) = $instr.maybe_peer_id() {
                    peer_id.to_string()
                } else {
                    $exec_ctx.run_parameters.current_peer_id.to_string()
                };

                $exec_ctx.set_errors(&e, peer_id, &$instr.to_string(), None);

                Err(e)
            }
            v => v,
        }
    }};
}
```

Reply:
But again, there is a lot of boilerplate code near each instruction.